[ContextParallel] add `_LoadBalancer` classes, and load-balance interface to Context Parallel APIs

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163806
* #163053
* #163617
* __->__ #161062

**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a rearrangement of the input tensor(s) over the
seq_dim before sharding and can be specified via an index tensor `rearrange` such
that Q[rearrange] is the balanced Q users want (i.e. `rearrange[i] == j` where `i` is the new
index of `Q[j]` in the balanced Q). An example is the `_generate_round_robin_indices()` added
in https://github.com/pytorch/pytorch/pull/155442. 

**New `_LoadBalancer` classes**
New `_LoadBalancer` class (defined in `torch/distributed/tensor/experimental/_load_balancer.py`)
provides one interface for defining load-balance behavior: `_generate_indices(self, restore: bool = False)`.

When `restore == False`, this method should output an index Tensor (namely `rearrange_idx`) such
that QKV will be transformed into Q' K' V' in a way that `Q'[i] == Q[rearrange_idx[i]]` (same applies
to K and V).

When `restore == True`, this method outputs an index Tensor (namely `restore_idx` such that
`Q'[restore_idx] == Q` (same applies to K and V).

**Impact**
2 public CP APIs and 1 private CP API is modified. This PR should be backward-compatible by:
- For uses w/ SDPA, existing users must be using the `context_parallel()` API which does not
take in the extra `load_balancer` argument and solely determines from the global var
`_cp_options.enable_load_balance`.
- For new users including who want to try `flex_attention()`, we require to use the new API
`_context_parallel_buffers` to explicitly shard the QKV input instead of using `context_parallel()`
because we no longer rely on TorchDispatchMode nor TorchFunctionMode for op replacement. And
we also require users to explicitly pass in a `load_balancer` argument if load-balancing is demanded.

**Load-Balance Behavior**
`context_parallel_unshard()`, and `create_cp_block_mask()` APIs now take an extra optional argument
`load_balancer`. This argument is optional because of backward compatibility but we require new users
to explicitly pass in a `load_balancer` if load-balancing is demanded:
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on).

`_context_parallel_buffers()` also takes the extra optional argument `load_balancer`, but the behavior
is slightly different from the other 2 APIs -- it doesn't branch on `_cp_options.enable_load_balance` :
- if `load_balancer == None`, no load-balancing will be performed
- otherwise, apply load-balancing using `load_balancer._generate_indices()` before sharding.

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* _context_parallel_buffers
* context_parallel_unshard
* create_cp_block_mask

The `_LoadBalancer` classes added are:
- `_LoadBalancer`: the abstract base class that provides “_generate_indices” interface index Tensor generation.
- `_HeadTailLoadBalancer`: Implements head-tail balancing logic.
- `_PerDocumentHeadTailLoadBalancer`: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

XilunWu committed:
[ContextParallel] add load-balance interface to context_parallel()

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP] [ContextParallel] add load-balance interface to context_parallel()"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP] [ContextParallel] add load-balance interface to context_parallel()"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP] [ContextParallel] add load-balance interface to context_parallel()"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP] [ContextParallel] add load-balance interface to context_parallel()"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

drisspg commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -104,6 +104,163 @@ class _ContextParallelGlobalVars:
+    def __init__(self) -> None:
+        pass
+
+    def generate_indices(self, restore: bool = False) -> Optional[torch.Tensor]:
```
we should figure out how to design this API such that it can also consume the BlockMask or at least parts. I want to  able to build something like https://en.wikipedia.org/wiki/Longest-processing-time-first_scheduling
from the information about the amount of work in the BlockMask

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fegin reviewed (approved):
Discussed offline, the two main issues are that the legacy `context_parallel()` shouldn't have a new load_balancer argument and we should move all the load balancer to a new file `attention_load_balancer.py`.

XilunWu commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -1201,6 +1520,7 @@ def context_parallel(
     buffers: Optional[list[torch.Tensor]] = None,
     buffer_seq_dims: Optional[list[int]] = None,
     no_restore_buffers: Optional[set[torch.Tensor]] = None,
+    load_balancer: Optional[LoadBalancer] = None,
```
offline sync with Chien-chin and decide to make the following change:
- keep the original context_parallel() signature. 
- add Optional[_LoadBalancer] argument to `_context_parallel_buffers` later will rename to `_context_parallel_shard`

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

XilunWu committed:
Update on "[ContextParallel] add load-balance interface to context_parallel()"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a shuffling of the input tensor(s) before sharding
and can be specified via an index tensor that marks the transition `pre_shuffle_idx`
`-> post_shuffle_idx`. An example is the `_generate_round_robin_indices` function added
in https://github.com/pytorch/pytorch/pull/155442. 

**Load-Balance Behavior**
`context_parallel`, `context_parallel_unshard`, and `create_cp_block_mask` APIs now
take extra optional argument `load_balancer`.
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on). For `context_parallel`, this is
performed before sharding the tensor; for `context_parallel_unshard` a reversed operation is
performed after unsharding.
- if `load_balancer` is a `LoadBalancer` object, its `generate_indices()` method is called to
generate the index shuffle Tensor (or its reverse).

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* context_parallel
* context_parallel_unshard
* create_cp_block_mask

The LoadBalancer classes added are:
- LoadBalancer: the abstract base class that provides “generate_indices” interface index Tensor generation.
- _DummyLoadBalancer: the LoadBalancer that implements “no-load-balancing” by having “generate_indices” return None.
- HeadTailLoadBalancer: Implements head-tail (round-robin) balancing logic.
- PerDocumentHeadTailLoadBalancer: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`

Impacted tests:
- `test_ring_flex_attention` added tests w/ load-balance. The `load_balancer` argument is always set to `None`.
- `test_ring_flex_attention_document_mask` added tests w/ load-balance. The `load_balancer` argument is `None` when load-balance is disabled, and assigned a `PerDocumentHeadTailLoadBalancer` object when load-balance is enabled.

Existing tests that have no impact (those tests are for SDPA and the `load_balancer` argument to APIs should be `None` for SDPA):
- `test_is_causal_behavior` 
- `test_ring_attention_native_transformer`
- `test_ring_attention_custom_transformer`
- `test_ring_attention_sdpa`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/experimental/_attention.py:
```diff
@@ -72,6 +72,312 @@ class _ContextParallelGlobalVars:
+class PerDocumentHeadTailLoadBalancer(LoadBalancer):
+    def __init__(
+        self,
+        seq_length_per_doc: Union[list[int], list[list[int]]],
```
The tensors are all created on device only. See: https://github.com/pytorch/pytorch/pull/161062/files#diff-079aa31d1e45ee129cb9e15298e1262345888e2ab99da917635488823cb2d5a1R267-R280

XilunWu committed:
Update on "[ContextParallel] add `_LoadBalancer` classes, and load-balance interface to Context Parallel APIs"


**Summary**
This PR provides an interface for users to specify how to load-balance the attention
input. The load-balance is essentially a rearrangement of the input tensor(s) over the
seq_dim before sharding and can be specified via an index tensor `rearrange` such
that Q[rearrange] is the balanced Q users want (i.e. `rearrange[i] == j` where `i` is the new
index of `Q[j]` in the balanced Q). An example is the `_generate_round_robin_indices()` added
in https://github.com/pytorch/pytorch/pull/155442. 

**New `_LoadBalancer` classes**
New `_LoadBalancer` class (defined in `torch/distributed/tensor/experimental/_load_balancer.py`)
provides one interface for defining load-balance behavior: `_generate_indices(self, restore: bool = False)`.

When `restore == False`, this method should output an index Tensor (namely `rearrange_idx`) such
that QKV will be transformed into Q' K' V' in a way that `Q'[i] == Q[rearrange_idx[i]]` (same applies
to K and V).

When `restore == True`, this method outputs an index Tensor (namely `restore_idx` such that
`Q'[restore_idx] == Q` (same applies to K and V).

**Impact**
2 public CP APIs and 1 private CP API is modified. This PR should be backward-compatible by:
- For uses w/ SDPA, existing users must be using the `context_parallel()` API which does not
take in the extra `load_balancer` argument and solely determines from the global var
`_cp_options.enable_load_balance`.
- For new users including who want to try `flex_attention()`, we require to use the new API
`_context_parallel_buffers` to explicitly shard the QKV input instead of using `context_parallel()`
because we no longer rely on TorchDispatchMode nor TorchFunctionMode for op replacement. And
we also require users to explicitly pass in a `load_balancer` argument if load-balancing is demanded.

**Load-Balance Behavior**
`context_parallel_unshard()`, and `create_cp_block_mask()` APIs now take an extra optional argument
`load_balancer`. This argument is optional because of backward compatibility but we require new users
to explicitly pass in a `load_balancer` if load-balancing is demanded:
- if `load_balancer == None` and `_cp_options.enable_load_balance == False`, CP performs
no load-balancing on input Tensors.
- if `load_balancer == None` and `_cp_options.enable_load_balance ==True`, CP performs
head-tail load-balancing (e.g. split a Tensor into 2*N chunks and first N are called head and
the rest are called tail. Place the first head chunk the last tail chunk on rank 0, and the second
head along with the second last tail chunk on rank 1, and so on).

`_context_parallel_buffers()` also takes the extra optional argument `load_balancer`, but the behavior
is slightly different from the other 2 APIs -- it doesn't branch on `_cp_options.enable_load_balance` :
- if `load_balancer == None`, no load-balancing will be performed
- otherwise, apply load-balancing using `load_balancer._generate_indices()` before sharding.

**Changes**
This PR moves the index Tensor generation logic into a set of LoadBalancer classes and
make LoadBalancer the common interface for Context Parallel APIs that leverages
load-balancing:
* _context_parallel_buffers
* context_parallel_unshard
* create_cp_block_mask

The `_LoadBalancer` classes added are:
- `_LoadBalancer`: the abstract base class that provides “_generate_indices” interface index Tensor generation.
- `_HeadTailLoadBalancer`: Implements head-tail balancing logic.
- `_PerDocumentHeadTailLoadBalancer`: Supports per-document head-tail balancing for batched sequences.

**Test**
`pytest test/distributed/tensor/test_attention.py`


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]
